src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -382,7 +382,13 @@ JvmtiExport::get_jvmti_interface(JavaVM *jvm, void **penv, jint version) {
   }
   if (Continuations::enabled()) {
     // Virtual threads support. There is a performance impact when VTMS transitions are enabled.
-    java_lang_VirtualThread::set_notify_jvmti_events(true);
+    if (!java_lang_VirtualThread ::notify_jvmti_events()) {
+      java_lang_VirtualThread::set_notify_jvmti_events(true);
+      if (Thread::current_or_null() != NULL) { // need this for attached threads only
+        ThreadInVMfromNative tiv(JavaThread::current());
+        java_lang_VirtualThread::init_static_notify_jvmti_events();
+      }
```

Comment:
Doesn't this logic mean that if the first pass through this code is made with an unattached thread, then that will prevent subsequent passes from calling `init_static_notify_jvmti_events`, even if the thread is attached. The reason is because `java_lang_VirtualThread::set_notify_jvmti_events(true);` will already have been done, so you won't pass the `if (!java_lang_VirtualThread ::notify_jvmti_events())` check.

Reply:
Enabling the `notify_jvmti_events` is an optimization to avoid having this notification overhead with JVMTI virtual thread mount state transitions when it is not needed.
We need to enable it only once and never disable it if enabled.
The first attempt to enable it is at startup if there was any agent loaded with command line options.
In such a case, the get_jvmti_interface() is called in a context of `AgentOnLoad()` in unattached thread.
It only sets the `java_lang_VirtualThread::set_notify_jvmti_events(true)` The `init_static_notify_jvmti_events()` is called from the `javaClasses_init()`.
The agents that are loaded into running VM are initialized with the `AgentOnAttach()`.
In this case, we can't rely on the `javaClasses_init()` and so, have to explicitly call the `init_static_notify_jvmti_events()`.
I feels like this can be simplified. I keep thinking about the best way to do it.
Probably, the pair <set_notify_jvmti_events,  init_static_notify_jvmti_events> can be replaced with just function.

Reply:
If `notify_jvmti_events()` is false, then you call `set_notify_jvmti_events(true)`, which means you will never enter the `if` block again. However, if the thread is not attached,  you do not call `init_static_notify_jvmti_events()`. What happens if later there is an attached thread that triggers this code? Is seem when that happens you should call `init_static_notify_jvmti_events()`, but won't because `notify_jvmti_events()` is true.

Reply:
I think you need a flag that tells you if `init_static_notify_jvmti_events()` has been called.

Reply:
A part of the initialization sequence we need know is:
```
create_vm()  {
  . . .
  // Launch -agentlib/-agentpath and converted -Xrun agents
  if (Arguments::init_agents_at_startup()) {
    create_vm_init_agents();  =>  {
       <loads all agents and calls AgentOnLoad entry points>  =>
        get_jvmti_interface() => set_notify_jvmti_events(true)
  }
  . . .
  init_globals() => javaClasses_init() => java_lang_VirtualThread::init_static_notify_jvmti_events()
```
The `create_vm_init_agents()` is called in the context of unattaching thread.
In this context a call to `java_lang_VirtualThread::init_static_notify_jvmti_events()` is guaranteed to happen after all the agents loaded at startup executed their `AgentOnLoad` entree points which make calls to `vm->GetEnv()` that transitively call `get_jvmti_interface()` and `java_lang_VirtualThread::set_notify_jvmti_events(true)`.

Reply:
If there are no command line agents, then on startup `vthread_notify_jvmti_events` is not set true. Because it is not true, when `javaClasses_init()` calls `init_static_notify_jvmti_events()`, it does nothing. The whole point of the code we are reviewing here is to make sure `init_static_notify_jvmti_events()` is called while `vthread_notify_jvmti_events == true` so it actually does something. However, the code here does not bother calling `init_static_notify_jvmti_events()` if the current thread is detached, but it does still set `vthread_notify_jvmti_events = true`. This means that if this code gets called a second time, this time with the current thread attached, it will not call `init_static_notify_jvmti_events()` due to `vthread_notify_jvmti_events == true`, but it seems it should be calling it.

What I believe to be the flaw here is that you call `set_notify_jvmti_events(true)` even if you don't call `init_static_notify_jvmti_events()`.

Reply:
> What I believe to be the flaw here is that you call `set_notify_jvmti_events(true)`
> even if you don't call `init_static_notify_jvmti_events()`.

This only happen in a detached thread case which can be only at startup.
It was implemented this way before my fix.
The javaClasses has to be partially initialized before any call to `init_static_notify_jvmti_events()`:
```
void javaClasses_init() {
  JavaClasses::compute_offsets();               <== This is needed
  JavaClasses::check_offsets();
  java_lang_VirtualThread::init_static_notify_jvmti_events();
  FilteredFieldsMap::initialize();  // must be done after computing offsets.
}
```

For detached thread case (which is at starup) it is guaranteed that the `init_static_notify_jvmti_events()` is unconditionally called later at initialization stage. Please, see the chain of calls:
`create_vm()  => init_globals() => javaClasses_init() => java_lang_VirtualThread::init_static_notify_jvmti_events()`.

Reply:
I've pushed an update.
Please, let me know of you are okay with that.